go/src/koding/kites/kloud/provider/softlayer/stack.go

```diff
@@ -0,0 +1,213 @@
+func (s *Stack) injectBootstrap(b *Bootstrap, guest map[string]interface{}) error {
+	bootstrapKey, err := strconv.Atoi(b.KeyID)
+	if err != nil {
+		return err
+	}
+
+	// Apply user ssh keys if provided, otherwise use the bootstrap key
+	keys := []int{bootstrapKey}
+
+	if userKeys, ok := guest["ssh_keys"].([]int); ok {
+		keys = append(keys, userKeys...)
+	}
+	guest["ssh_keys"] = keys
+
+	guest["image"] = "UBUNTU_14_64"
+	// NOTE: Should we allow for user override here? Do all images work with klient?
+	if userImage, ok := guest["image"]; ok {
+		guest["image"] = userImage
+	}
+
+	return nil
+}
+
+// injectKiteKeys creates a kite key for each virtual guest, and sets up
+// the runtime injection of said key using cloud-init/user_data
+func (s *Stack) injectKiteKeys(guest map[string]interface{}, name string) error {
+	s.Builder.InterpolateField(guest, name, "user_data")
+
+	kiteKey, err := s.BuildKiteKey(name, s.Req.Username)
+	if err != nil {
+		return err
+	}
+
+	// Map the generated kite key to the guest via cloud-init config
+	cloudInitConfig := &userdata.CloudInitConfig{
+		Username: s.Req.Username,
+		Groups:   []string{"sudo"},
+		Hostname: s.Req.Username,
+		KiteKey:  kiteKey,
+	}
+
+	// If user provided user data, associate that with our cloud init config
+	if ud, ok := guest["user_data"].(string); ok {
+		cloudInitConfig.UserData = ud
+	}
+
+	// Create cloud init and associate with guest
+	cloudInit, err := s.Session.Userdata.Create(cloudInitConfig)
+	if err != nil {
+		return err
+	}
+	guest["user_data"] = string(cloudInit)
+
+	return nil
+}
+
+// ApplyTemplate injects bootstrap resources into a SoftLayer stack,
+// and is responsible for ensuring each new Softlayer instance will
+// connect to Koding upon start.
+func (s *Stack) ApplyTemplate(c *stack.Credential) (*stack.Template, error) {
+	t := s.Builder.Template
+
+	var resource struct {
+		VirtualGuest map[string]map[string]interface{} `hcl:"softlayer_virtual_guest"`
+	}
+
+	// Decode the user provided stack template, searching for
+	// virtual guest definitions
+	if err := s.Builder.Template.DecodeResource(&resource); err != nil {
+		return nil, err
+	}
+
+	// If user hasn't specified a virtual guest definitions
+	// in their stack, then we have nothing else to do
+	if len(resource.VirtualGuest) == 0 {
+		return nil, errors.New("No softlayer_virtual_guest resources were defined in stack.")
+	}
+
+	bootstrap := c.Bootstrap.(*Bootstrap)
+
+	// Iterate through n softlayer_virtual_guest definitions, filling in
+	// configuration gaps with bootstrap, and setting up kite keys with
+	// cloud-init for each.
+	for name, guest := range resource.VirtualGuest {
+		if err := s.injectBootstrap(bootstrap, guest); err != nil {
+			return nil, err
+		}
+
+		if err := s.injectKiteKeys(guest, name); err != nil {
+			return nil, err
+		}
+	}
+
+	t.Resource["softlayer_virtual_guest"] = resource.VirtualGuest
+
+	// Don't show secret things in frontend logs to the user
```

Comment:
In case of Softlayer probably it's not that important. The softlayer-go library misuses Basic Auth, instead of adding a header to a HTTP request it adds auth info to the hostname:

https://github.com/maximilien/softlayer-go/blob/master/client/http_client.go#L69

So if request times out, go's `net/http` library is logging the whole hostname with `*net.OpError` errors. So, until the lib is fixed, the credentials will leak. Just FYI.